File: packages/clawde-sdk/src/circuit-breaker.ts
import { CircuitBreakerOptions } from './types'
import { ClawdeVerdictError } from './errors'

export class CircuitBreaker {
  private client: any

  constructor(client: any) {
    this.client = client
  }

  public wrap<T>(
    toolName: string,
    options: CircuitBreakerOptions = {}
  ): (fn: () => Promise<T>) => Promise<T> {
    const failOpen = options.failOpen ?? false

    return async (fn: () => Promise<T>): Promise<T> => {
      // 1. Pre-Check: Context & pre-flight budget validation
      try {
        const context = await this.client.resolveContext()
        // If maxCostUsd is specified, check against current budget limit or remaining
        if (options.maxCostUsd !== undefined) {
          // Pre-flight check
          const budget = await this.client.checkBudget('default', 1)
          if (!budget.allowed) {
            throw new ClawdeVerdictError('kill', `Circuit breaker tripped for tool '${toolName}': budget exceeded. Remaining: $${budget.remaining_usd}`)
          }
        }
      } catch (err: any) {
        if (!failOpen) {
          throw err
        }
        // If failOpen is true, we log the error and continue execution
        if (process.env.INTUTIC_DEBUG === 'true') {
          console.warn(`[Clawde SDK] Circuit breaker pre-check failed (failing open): ${err.message}`)
        }
      }

      // 2. Execute the function
      try {
        const result = await fn()
        
        // 3. Post-Check: If the result is a ChatResponse (or has verdict headers), inspect it
        if (result && typeof result === 'object') {
          const res = result as any
          if (res.verdict === 'kill') {
            throw new ClawdeVerdictError('kill', `Execution blocked by governance policy (Verdict: KILL)`)
          }
        }
        
        return result;
      } catch (err: any) {
        if (err instanceof ClawdeVerdictError && !failOpen) {
          throw err
        }
        if (!failOpen) {
          throw err
        }
        if (process.env.INTUTIC_DEBUG === 'true') {
          console.warn(`[Clawde SDK] Circuit breaker execution failed (failing open): ${err.message}`)
        }
        return null as any // Fail open returns null or empty
      }
    }
  }
}

File: packages/intutic-clawde/intutic_clawde/circuit_breaker.py
import os
from typing import Callable, Any, Optional, TypeVar
from .errors import ClawdeVerdictError

T = TypeVar('T')

class CircuitBreaker:
    def __init__(self, client: Any):
        self.client = client

    def wrap(self, tool_name: str, max_cost_usd: Optional[float] = None, fail_open: bool = False) -> Callable[[Callable[[], T]], T]:
        def decorator(fn: Callable[[], T]) -> T:
            # 1. Pre-Check: Context & pre-flight budget validation
            try:
                if max_cost_usd is not None:
                    budget = self.client.check_budget("default", 1)
                    if not budget.get("allowed", True):
                        raise ClawdeVerdictError(
                            "kill", 
                            f"Circuit breaker tripped for tool '{tool_name}': budget exceeded. Remaining: ${budget.get('remaining_usd')}"
                        )
            except Exception as e:
                if not fail_open:
                    raise e
                if os.environ.get("INTUTIC_DEBUG") == "true":
                    print(f"[Clawde SDK] Circuit breaker pre-check failed (failing open): {str(e)}")
                    
            # 2. Execute the function
            try:
                result = fn()
                
                # 3. Post-Check: If the result is a dict with verdict key, inspect it
                if isinstance(result, dict) and result.get("verdict") == "kill":
                    raise ClawdeVerdictError("kill", "Execution blocked by governance policy (Verdict: KILL)")
                    
                return result
            except Exception as e:
                if isinstance(e, ClawdeVerdictError) and not fail_open:
                    raise e
                if not fail_open:
                    raise e
                if os.environ.get("INTUTIC_DEBUG") == "true":
                    print(f"[Clawde SDK] Circuit breaker execution failed (failing open): {str(e)}")
                return None # type: ignore
        return decorator # type: ignore
